<img style="float: right" height="50%" width="50%" src="img/logo.png">

### Abundant Bacterial Species in YOUR Beyond Bacteria Sample

All four Beyond Bacteria participants had very different microbiome profiles
when looking at the top ten genera detected in each participant. The microbiome
composition of each participant provided intriguing bits of food for thought.
Disclaimer: science is not at the stage to confidently assess human health
status based on the microbiome, and the American Gut is not a medical
diagnostic. The microbiome could also follow the old adage that too much of a
good thing is a bad thing, and not enough studies have been done to determine
whether certain abundance thresholds of "good" bacteria become harmful. Only
time and well-designed, peer-reviewed research will tell us how useful
microbiome composition can be as a proxy for predicting human health status.

You can find the specific top taxa in your sample
[here](http://nbviewer.ipython.org/github/biocore/American-Gut/blob/master/ipynb/beyond_bacteria.ipynb#000011924_taxa).

Seven of the top ten species detected in your gut microbiome (000011924) were
*Lactobacillus* species, and an additional two were *Bifidobacterium* species, both
of which are generally typified as the probiotic strains. As discussed, many of
these species have been tested for their probiotic potential in a number of
contexts and have a wide range of purported effects, from a reduction in
adiposity to a reduction in allergy related inflammation. The large proportion
of *Lactobacillus spp.* in your gut is definitely curious, though the cause for
this cannot be determined from the lifestyle information you provided. In
general, you appear to live a healthy lifestyle, eating a variety of vegetables
and exercising daily. Nevertheless, you are slightly underweight on the BMI
scale. You did use antibiotics in the year prior to collecting your sample. An
attempt to prevent or correct antibiotic-induced microbiome dysbiosis via the
ingestion of probiotics could potentially explain the large amount of
probiotic-like species in your gut, but without this specific information, this
potential explanation remains conjecture.

### Notable Metabolic Pathways in YOUR Beyond Bacteria Sample

The most abundant pathways were similar across the four individuals, despite
taxonomic differences. This is not unprecedented; the Human Microbiome Project
also large variations in taxonomy, while pathway abundance remained relatively
stable ([ref](http://www.ncbi.nlm.nih.gov/pubmed/22699609)). Unsurprisingly, we
find common housekeeping genes, like DNA biosynthesis and carbon metabolism,
across all four samples. We also find less common pathways. Again, we must
issue a disclaimer: metagenomics is still in its infancy. Few samples have been
analyzed in this way. Therefore, making health inferences based on metagenomic
findings would be very premature at this time.

You can find the specific top pathways in your sample
[here](http://nbviewer.ipython.org/github/biocore/American-Gut/blob/master/ipynb/beyond_bacteria.ipynb#000011924_pathway).

Your metagenome (000011942) follows a general pattern of core pathways. Your
sample lacks pathways associated with any type of aromatic degradation,
including phenylalanine. Aromatics are rigid, stable carbon rings. Traditional
physiology places aromatic degradation in the human liver, although some
microbes include this capacity.  The bacteria do not encode genes to break down
glycosaminoglycan. Glycosaminoglycan are large, branched sugars found on the
surface of host cells. In their native form, glycosaminoglycans can promote
infection. Bacterial breakdown of glycosaminoglycans can promote an immune
response, leading to faster clearance
([ref](http://www.ncbi.nlm.nih.gov/pubmed/25122767)). Your metagenome also
lacks some pathways involved in energy metabolism, including enzymes to convert
between hydrogen sulfide (H2S) and sulfate (SO42-). Hydrogen sulfide has a
distinct rotten egg aroma and is commonly found in flatulence.
